subsys/net/lib/http/http_server.c

```diff
@@ -0,0 +1,1235 @@
+		ctx_client->offset = 0;
+	}
+
+	return 0;
+}
+
+int handle_http_done(struct http_server_ctx *ctx_server, struct http_client_ctx *ctx_client,
+		     int client_index)
+{
+	LOG_DBG("HTTP_SERVER_DONE_STATE");
+
+	close_client_connection(ctx_server, client_index);
+
+	return -errno;
+}
+
+int handle_http_frame_headers(struct http_client_ctx *ctx_client)
+{
+	LOG_DBG("HTTP_SERVER_FRAME_HEADERS");
+
+	unsigned char response_headers_frame[16];
+
+	struct http_frame *frame = &ctx_client->current_frame;
+
+	print_http_frames(ctx_client);
+
+	if (ctx_client->offset < frame->length) {
+		return -EAGAIN;
+	}
+
+	const char *method;
+	const char *path;
+
+	if (!has_upgrade_header) {
+		enum http_method method_h2c = parser.method;
+
+		method = http_method_str(method_h2c);
+		path = url_buffer;
+	} else {
+		method = http_hpack_parse_header(ctx_client, HTTP_SERVER_HPACK_METHOD);
+		path = http_hpack_parse_header(ctx_client, HTTP_SERVER_HPACK_PATH);
+	}
+
+	if (strcmp(method, "GET") == 0 && strcmp(path, "/") == 0) {
+
+		generate_response_headers_frame(response_headers_frame, frame->stream_identifier,
+						HTTP_SERVER_HPACK_STATUS_2OO);
+
+		size_t ret = sendall(ctx_client->client_fd, response_headers_frame,
+				     sizeof(response_headers_frame));
+		if (ret < 0) {
+			LOG_ERR("ERROR writing to socket");
+			return -errno;
+		}
+
+		size_t content_size = sizeof(content_200);
+
+		send_data(ctx_client->client_fd, content_200, content_size, HTTP_SERVER_DATA_FRAME,
+			  HTTP_SERVER_FLAG_END_STREAM, frame->stream_identifier);
+	} else {
+		generate_response_headers_frame(response_headers_frame, frame->stream_identifier,
+						HTTP_SERVER_HPACK_STATUS_4O4);
+
+		size_t ret = sendall(ctx_client->client_fd, response_headers_frame,
+				     sizeof(response_headers_frame));
+		if (ret < 0) {
+			LOG_ERR("ERROR writing to socket");
+			return -errno;
+		}
+
+		size_t content_size = sizeof(content_404);
+
+		send_data(ctx_client->client_fd, content_404, content_size, HTTP_SERVER_DATA_FRAME,
+			  HTTP_SERVER_FLAG_END_STREAM, frame->stream_identifier);
+	}
+
+	ctx_client->server_state = HTTP_SERVER_FRAME_HEADER_STATE;
+
+	int bytes_consumed = ctx_client->current_frame.length;
+
+	ctx_client->offset -= bytes_consumed;
+	memmove(ctx_client->buffer, ctx_client->buffer + bytes_consumed, ctx_client->offset);
+
+	return 0;
+}
+
+int handle_http_frame_priority(struct http_client_ctx *ctx_client)
+{
+	LOG_DBG("HTTP_SERVER_FRAME_PRIORITY_STATE");
+
+	struct http_frame *frame = &ctx_client->current_frame;
+
+	print_http_frames(ctx_client);
+
+	if (ctx_client->offset < frame->length) {
+		return -EAGAIN;
+	}
+
+	int bytes_consumed = ctx_client->current_frame.length;
+
+	ctx_client->offset -= bytes_consumed;
+	memmove(ctx_client->buffer, ctx_client->buffer + bytes_consumed, ctx_client->offset);
+
+	ctx_client->server_state = HTTP_SERVER_FRAME_HEADER_STATE;
+
+	return 0;
+}
+
+int handle_http_frame_continuation(struct http_client_ctx *ctx_client)
```

Comment:
CONTINUATION frame should be handled similarily to HEADERS frame - it carries additional HPACK headers.